### PR TITLE
Bluetooth: samples: Add ISO connected benchmark sample

### PR DESCRIPTION
**Pull request by Thalley**
Add a sample that measures packet loss in a connected
ISO setup.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Was originally https://github.com/zephyrproject-rtos/zephyr/pull/35602 but then reverted in https://github.com/zephyrproject-rtos/zephyr/pull/36750 due to a CI error.

### CODE REVIEW

File: samples/bluetooth/iso_connected_benchmark/sample.yaml
Comment (nashif) [inline samples/bluetooth/iso_connected_benchmark/sample.yaml:7]: why do you need this?

File: samples/bluetooth/iso_connected_benchmark/sample.yaml
Comment (nashif) [inline samples/bluetooth/iso_connected_benchmark/sample.yaml:6]: this is a benchmark so why build_only?

**Review by nashif — CHANGES_REQUESTED** [review]

File: samples/bluetooth/iso_connected_benchmark/sample.yaml
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/sample.yaml]: I don't; removed.

File: samples/bluetooth/iso_connected_benchmark/sample.yaml
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/sample.yaml:6]: The benchmark sample application doesn't run autonomously, so we can only build it.

File: samples/bluetooth/iso_connected_benchmark/README.rst
Comment (asbjornsabo) [inline samples/bluetooth/iso_connected_benchmark/README.rst:7]: in terms "of"

Also unsure if "quality" is a good word to use here.  But as long as it is qualified with the "packet and sync loss", it should be OK.
An alternative might be something along the lines of :
"The ISO Connected Channels Benchmark sample measures and reports packet loss and sync loss in connected ISO channels."

File: samples/bluetooth/iso_connected_benchmark/README.rst
Comment (asbjornsabo) [inline samples/bluetooth/iso_connected_benchmark/README.rst:1]: broadcast -> "connected" (or "connected_channels"9

File: samples/bluetooth/iso_connected_benchmark/README.rst
Comment (asbjornsabo) [inline samples/bluetooth/iso_connected_benchmark/README.rst:16]: mode?  Whether the application is in the central role or the peripheral role?

Not used as both at the same time, I guess?

The application can be used as either a central or a peripheral, and can easily switch between these roles.

File: samples/bluetooth/iso_connected_benchmark/README.rst
Comment (asbjornsabo) [inline samples/bluetooth/iso_connected_benchmark/README.rst:47]: How?

File: samples/bluetooth/iso_connected_benchmark/README.rst
Comment (asbjornsabo) [inline samples/bluetooth/iso_connected_benchmark/README.rst:14]: and measures sync loss and packet loss

File: samples/bluetooth/iso_connected_benchmark/README.rst
Comment (asbjornsabo) [inline samples/bluetooth/iso_connected_benchmark/README.rst:82]: broadcaster (2x)?  BIG?

**Review by asbjornsabo — COMMENTED** [review]
Preliminary comments (a bit nit-picking, I know, I think the README can be improved)

File: samples/bluetooth/iso_connected_benchmark/README.rst
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/README.rst:16]: Correct.

File: samples/bluetooth/iso_connected_benchmark/README.rst
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/README.rst:22]: The sample just needs to run on a single board. It doesn't matter if the peer is running the same sample or even running a consumer product.

File: samples/bluetooth/iso_connected_benchmark/README.rst
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/README.rst:47]: Added "using the console"

File: samples/bluetooth/iso_connected_benchmark/README.rst
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/README.rst:14]: Already updated as you suggested

File: samples/bluetooth/iso_connected_benchmark/README.rst
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/README.rst:82]: Fixed

File: samples/bluetooth/iso_connected_benchmark/README.rst
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/README.rst:22]: Updated

**Review by nashif — APPROVED** [review]

File: samples/bluetooth/iso_connected_benchmark/src/main.c
Comment (ppryga) [inline samples/bluetooth/iso_connected_benchmark/src/main.c:877]: Remove empty line

File: samples/bluetooth/iso_connected_benchmark/src/main.c
Comment (ppryga) [inline samples/bluetooth/iso_connected_benchmark/src/main.c:900]: nit: the name of the function looks like one from unit tests. Maybe change to run_benchmark_for_peripheral?

**Review by ppryga-nordic — COMMENTED** [review]
I know that the sample was already approved in PR that was reverted, so requesting changes may be frustrating to you. I put few comments, they are nit-picking. Generally the code looks ok.

File: samples/bluetooth/iso_connected_benchmark/README.rst
Comment (asbjornsabo) [reply samples/bluetooth/iso_connected_benchmark/README.rst]: Just wondering whether it would be better to actually use the word "role", rather than "mode", when talking about switching to the other role.

File: samples/bluetooth/iso_connected_benchmark/src/main.c
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/src/main.c:41]: Changed to use one of the "standard" adv parameters.

File: samples/bluetooth/iso_connected_benchmark/src/main.c
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/src/main.c:985]: We cannot use printk in all places (possible these in the setup phase), as the time to print negatively affects the test. 

Likewise, we cannot use LOG everywhere as that does not work well with applications that expect input from the user. 

If you like, I can update the PR to use printk for everything before we actually start the test?

File: samples/bluetooth/iso_connected_benchmark/src/main.c
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/src/main.c:983]: Correct, leftover from the broadcast test that works slightly different :)

File: samples/bluetooth/iso_connected_benchmark/src/main.c
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/src/main.c:101]: Changed all occurrences of `console_getchar()` to `tolower(console_getchar())` so everything is uncapitalized

File: samples/bluetooth/iso_connected_benchmark/src/main.c
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/src/main.c:454]: Added a TODO. I'd rather create that in a separate PR that is unrelated to a sample application.

File: samples/bluetooth/iso_connected_benchmark/src/main.c
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/src/main.c:478]: see https://github.com/zephyrproject-rtos/zephyr/pull/36752#discussion_r668522110

File: samples/bluetooth/iso_connected_benchmark/src/main.c
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/src/main.c:502]: See https://github.com/zephyrproject-rtos/zephyr/pull/36752#discussion_r668522110

File: samples/bluetooth/iso_connected_benchmark/src/main.c
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/src/main.c:514]: Nope, but `strtoul` returns a uint64_t, so figured it was better to ensure that whatever the user entered would have have any implicit type casting. Alternatively I would need to check the value range before casting to a uint8_t and then check the value again.

File: samples/bluetooth/iso_connected_benchmark/src/main.c
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/src/main.c:553]: See https://github.com/zephyrproject-rtos/zephyr/pull/36752#discussion_r668522110

File: samples/bluetooth/iso_connected_benchmark/src/main.c
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/src/main.c:813]: Simply renamed to `run_{central/peripheral}`

File: samples/bluetooth/iso_connected_benchmark/src/main.c
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/src/main.c:900]: See https://github.com/zephyrproject-rtos/zephyr/pull/36752#discussion_r668525632

File: samples/bluetooth/iso_connected_benchmark/src/main.c
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/src/main.c:783]: The central was missing some proper cleanup. I've updated the way cleanup is performed, and using the same cleanup for both central and peripheral.

**Comment by Thalley** [comment]
@ppryga thanks for the review, I think I've fixed all of your comments. I'll apply the same changed to the broadcast PR (https://github.com/zephyrproject-rtos/zephyr/pull/35178)

File: samples/bluetooth/iso_connected_benchmark/src/main.c
Comment (asbjornsabo) [inline samples/bluetooth/iso_connected_benchmark/src/main.c:866]: "Failed to created CIS"?

File: samples/bluetooth/iso_connected_benchmark/src/main.c
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/src/main.c:200]: This is intentional. We log for each 100 of total packets (this may work poorly for very low or very high intervals), but that number is unrelated to the last 1000 packets. 

It is in theory possible to create a formula to make both of these numbers be relative to the number of CIS and the interval, but that's a thing for the future if ever.

File: samples/bluetooth/iso_connected_benchmark/src/main.c
Comment (Thalley) [reply samples/bluetooth/iso_connected_benchmark/src/main.c:204]: Yes. Added comment.

**Review by ppryga-nordic — APPROVED** [review]

**Comment by cfriedt** [comment]
@Thalley - is this one ready to be merged? Are comments resolved?

**Comment by Thalley** [comment]
@cfriedt I'd like to get an approval from @asbjornsabo before we merge, so he can verify that his comments were resolved

**Review by asbjornsabo — APPROVED** [review]
My comments have been resolved to my satisfaction

**Pull request merged**
